File: agent.py
from py_client.consumer import Consumer
from py_client.producer import Producer


class Agent:

    def __init__(self, producer: Producer, consumer: Consumer):
        self.producer = producer
        self.consumer = consumer

    def __enter__(self):
      return self

    def __exit__(self, exc_type, exc_value, exc_traceback):
        self.close()

    def close(self):
        self.producer.client.close()
        self.consumer.client.close()
    
    @staticmethod
    def of(host: str, port: int, client_id: str):
        producer = Producer(host, port, client_id)
        consumer = Consumer(host, port, client_id)
        
        return Agent(producer, consumer)

File: message_broker/app/src/main/java/capstone/design/client/Agent.java
package capstone.design.client;

public class Agent implements AutoCloseable {

    private final Producer producer;
    private final Consumer consumer;
    
    // constructor =====
    private Agent(Producer producer, Consumer consumer) {
        this.producer = producer;
        this.consumer = consumer;
    }

    // factory method =====
    public static Agent of(String host, int port, String clientId) throws Exception {
        return new Agent(
            new Producer(host, port, clientId),
            new Consumer(host, port, clientId)
        );
    }

    public static Agent of(Producer producer, Consumer consumer) {
        return new Agent(producer, consumer); 
    }

    // getter =====
    public Producer producer() { return producer; }
    public Consumer consumer() { return consumer; }

    @Override
    public void close() throws Exception {
        producer.close();
        consumer.close();
    }
}
